Add Stable Diffusion 3.5 Transformer as Model #332

ddilbazTT opened the pull request:
### Ticket
#332 

### Problem description
Stable Diffusion 3.5 is too big to add. It will eventually crash with `leaked semaphore` error. So, I am simply adding a Transformer Pass through SD3Transformer2DModel, which is the backbone of StableDiffusion 3.5.

### What's changed
Added licensing material and StableDiffusion 3.5 test file.

### Checklist
- [X] Added StableDiffusion 3.5 medium to e2e execute workflow, large to e2e compile workflow
- [X] Added StableDiffusion 3.5 medium and large to op-by-op nightly workflow
- [X] Created `tests/models/stable_diffusion/test_stable_diffusion_transformer.py`
- [X] Added "LICENSE.txt" and "NOTICE.txt" as per Stable Diffusion access requirements

kmabeeTT commented on .github/workflows/run-op-by-op-model-tests-nightly.yml:
```diff
@@ -163,6 +163,12 @@ jobs:
+          },
+          {
+            runs-on: wormhole_b0, name: "stable-diffusion-v3.5", tests: "
+              tests/models/stable_diffusion/test_stable_diffusion_transformer.py::test_stable_diffusion_transformer[op_by_op_torch-stable-diffusion-v3.5-medium-eval]
```
small thing and I apologize for having to mention it, but medium test is redundant here, shouldn't be in op-by-op nightly if in nightly full-model-execute, should be in op-by-op weekly.

kmabeeTT replied:
There is an additional change to make in this PR too: Now that I just merged the change https://github.com/tenstorrent/tt-torch/commit/48c2ca89c8f30ee85d6b4766b6a77abbdec03671 a few minutes ago, we want to add all `model_group=red` tests to full model execute nightly list (16 examples in that commit, please reference them) even that are not yet running e2e because stakeholders want these models' status reported on superset dashboards.  You can use the new function, something like this, right before `ThisTester` and right after finishing constructing `cc`:

```
    skip_full_eval_test(
        record_property,
        cc,
        model_name,
        bringup_status="FAILED_RUNTIME",
        reason="Model is too large to fit on single device during execution.",
        model_group=model_group,
        model_name_filter="stabilityai/stable-diffusion-3.5-large",
    )
```

kmabeeTT reviewed (approved):
looks good, there is one thing that could be changed (redundant test) but marking approved

AleksKnezevic commented on tests/models/stable_diffusion/NOTICE.txt:
```diff
@@ -0,0 +1,3 @@
+This repository includes Stability AI Model.
```
Does it?

AleksKnezevic commented on tests/models/stable_diffusion/test_stable_diffusion_transformer.py:
```diff
@@ -0,0 +1,116 @@
+    ids=["op_by_op_stablehlo", "op_by_op_torch", "full"],
+)
+def test_stable_diffusion_transformer(record_property, model_info, mode, op_by_op):
+    torch.set_num_threads(1)
```
What's this for?

ddilbazTT replied:
Sorry I forgot to remove it. [In one discussion thread,](https://github.com/apple/ml-stable-diffusion/issues/8#issuecomment-2639390262) some users were recommending to set `num_threads` to 1 when hitting a "leaked semaphore" crash. I will remove it.

AleksKnezevic commented on tests/models/stable_diffusion/test_stable_diffusion_transformer.py:
```diff
@@ -0,0 +1,116 @@
+
+
+model_info_list = [
+    ("stable-diffusion-v3.5-medium", "stabilityai/stable-diffusion-3.5-medium"),
```
Do you have an e2e test like we did for 1.3? If it's functional you may as well add it (at least to weekly if it takes too long). If it's not functional, I wouldn't worry about it at this time.

ddilbazTT replied:
So it runs op-by-op until it hits the leaked semaphore issue. But then I'm curious if there's ever something we can do on our end to not hit it. We agreed it is related to memory usage spike, but will we eventually have a chance to have more memory? Should I add the tests even though they do not run right now?

AleksKnezevic replied:
I think it's ok like this for now.

ddilbazTT commented on tests/models/stable_diffusion/LICENSE.txt:
```diff
@@ -0,0 +1,25 @@
+STABILITY AI COMMUNITY LICENSE AGREEMENT Last Updated: July 5, 2024
```
So this is in the license agreement:

> IV. GENERAL TERMS
> 
> Your Research, Non-Commercial, and Commercial License(s) under this Agreement are subject to the following terms. a. Distribution & Attribution. If You distribute or make available the Stability AI Materials or a Derivative Work to a third party, or a product or service that uses any portion of them, You shall: (i) provide a copy of this Agreement to that third party, (ii) retain the following attribution notice within a "Notice" text file distributed as a part of such copies: "This Stability AI Model is licensed under the Stability AI Community License, Copyright © Stability AI Ltd. All Rights Reserved", and (iii) prominently display "Powered by Stability AI" on a related website, user interface, blogpost, about page, or product documentation. 

Since we're using their pretrained model files, I think this applies.

ddilbazTT replied:
[License link for reference](https://huggingface.co/stabilityai/stable-diffusion-3.5-medium/blob/main/LICENSE.md)

AleksKnezevic replied:
We're not distributing the work in any way. If we were hosting the weights I think this would apply.

ddilbazTT commented:
@AleksKnezevic @kmabeeTT I will close this PR for now because since opening, I did a deep dive into the memory error. Doing some memory cleanup in pass_pipeline ended up helping and we can run SD3.5 with memory optimized architecture, medium version op-by-op. I will just launch some tests and add to workflows based on it. Regardless, I will keep the transformer test. I already addressed most comments in this PR and hopefully, I'll put it up again soon.

kmabeeTT commented:
> @AleksKnezevic @kmabeeTT I will close this PR for now because since opening, I did a deep dive into the memory error. Doing some memory cleanup in pass_pipeline ended up helping and we can run SD3.5 with memory optimized architecture, medium version op-by-op. I will just launch some tests and add to workflows based on it. Regardless, I will keep the transformer test. I already addressed most comments in this PR and hopefully, I'll put it up again soon.

Nice! Great job!   Also, some added info - I was just guessing in my example above, if you have model that won't run in full-model execute, adjust this as needed with a proper reason and bringup_status 

```
        bringup_status="FAILED_RUNTIME",
        reason="Model is too large to fit on single device during execution.",
```

available:

```
tests/utils.py:        bringup_status: The bringup status of the test (FAILED_FE_COMPILATION, FAILED_TTMLIR_COMPILATION, FAILED_RUNTIME, INCORRECT_RESULT, PASSED)
```
